eng/scripts/Cadl-Project-Sync.ps1

```diff
@@ -0,0 +1,121 @@
+[CmdletBinding()]
+param (
+    [Parameter(Position=0)]
+    [ValidateNotNullOrEmpty()]
+    [string] $ProjectDirectory
+)
+
+$ErrorActionPreference = "Stop"
+. $PSScriptRoot/../common/scripts/Helpers/PSModule-Helpers.ps1
+Install-ModuleIfNotInstalled "powershell-yaml" "0.4.1" | Import-Module
+$sparseCheckoutFile = ".git/info/sparse-checkout"
+
+function AddSparseCheckoutPath([string]$subDirectory) {
+    if (!(Test-Path $sparseCheckoutFile) -or !((Get-Content $sparseCheckoutFile).Contains($subDirectory))) {
+        Write-Output $subDirectory >> .git/info/sparse-checkout
+    }
+}
+
+function CopySpecToProjectIfNeeded([string]$specCloneRoot, [string]$mainSpecDir, [string]$dest, [string[]]$specAdditionalSubDirectories) {
+    $source = "$specCloneRoot/$mainSpecDir"
+    Write-Host "Copying spec from $source"
+    Copy-Item -Path $source -Destination $dest -Recurse -Force
+    foreach($additionalDir in $specAdditionalSubDirectories)
+    {
+        $source = "$specCloneRoot/$additionalDir"
+        Write-Host "Copying spec from $source"
+        Copy-Item -Path $source -Destination $dest -Recurse -Force
+    }
+}
+
+function UpdateSparseCheckoutFile([string]$mainSpecDir, [string[]]$specAdditionalSubDirectories) {
+    AddSparseCheckoutPath $mainSpecDir
+    foreach($subDir in $specAdditionalSubDirectories)
+    {
+        AddSparseCheckoutPath $subDir
+    }
+}
+
+function GetGitRemoteValue([string]$repo) {
+    Push-Location $ProjectDirectory
+    $result = ""
+    try {
+        $gitRemotes = (git remote -v)
+        foreach ($remote in $gitRemotes)
+        {
+            if ($remote.StartsWith("origin")) {
+                if ($remote -match 'https://github.com/\S+[\.git]') {
+                    $result = "git@github.com:$repo.git"
+                    #$result = "https://github.com/$repo.git"
+                    break
+                } elseif ($remote -match "git@github.com:(\S+)\.git"){
+                    $result = "git@github.com:$repo.git"
+                    break
+                } else {
+                    throw "Unknown git remote format found: $remote"
+                }
+            }
+        }
+    }
+    finally {
+        Pop-Location
+    }
+
+    return $result
+}
+
+function InitializeSparseGitClone([string]$repo) {
+    git clone --no-checkout --filter=tree:0 $repo .
```

Comment:
In order to get this to work for the private repo we are going to either need to use a PAT from our bot account or perhaps figure out how to use the credentials from the agent. @benbp did you ever do any investigation with the persistCredentials option https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/steps-checkout?view=azure-pipelines while doing your sparse checkout work?

Reply:
I think the approach currently used in test proxy where we filter on the PAT existing and change the repo url accordingly could work here as well. I recall initially trying to get pipeline sparse checkout with persist credentials working, but the format/encoding of the token was set up in some way where I couldn't figure out how to consume it. Some working examples may have been posted [here](https://github.com/Azure/azure-sdk-tools/issues/3222#issuecomment-1111116452).